PorousFlowCount added.

WilkAndy opened the pull request:
Refs #8901

Here i have used an ElementalUserObject, PorousFlowCount, to store some info that i need in my Kernel.  Could someone review this quickly and tell me whether my usage of an ElementalUserObject is reasonable?  The rest of the code can be ignored for now.

permcody commented on modules/porous_flow/src/kernels/PorousFlowDarcyBase.C:
```diff
@@ -93,22 +100,53 @@ PorousFlowDarcyBase::computeQpResidual()
+    }
+  }
+  if (_pf_count_used)
+    _pf_count->setProtoFlux(_proto_flux);
```
This seems really dangerous to me and is subject to breakage. In general in MOOSE you shouldn't be trying to shove data into other objects from some other object objects. With threading and decomposition, it's just a bad idea. In fact, our UserObject system, which is usually the extra storage dumping ground, simply doesn't allow this. I suspect you might be using ElementalAux to bypass that restriction.

Without really digging through all of this code, have you considered storing information right inside the kernel? Honestly that's not ideal either but is better then pushing information into another object in my opinion. Ideally UserObject-based solution would be my preferred method if possible, but I don't think that'll work particularly well since you are trying to track occurrences of something happening in a Kernel.

WilkAndy replied:
Yes, i'd like to store this in the Kernel, but i don't know how to do it.  At the start of each timestep i need to set up and zero _num_upwinds, etc, as std::map, presumably, but will that information persist and be able to be modified throughout each nonlinear iteration process?  I tried this but it didn't work - if you get time, can you give me a couple of clues (if not, i'll probably work it out myself over the next couple of days).

WilkAndy replied:
Oh, and i forgot to say "thanks heaps" for reviewing this rather messy code - probably saved me tonnes of effort!

WilkAndy replied:
Didn't you say std::map was super slow, by the way?  I know premature optimisation is bad, but this Kernel is the main Kernel that will appear in all PorousFlow simulations.

permcody replied:
Kernels have the `timetepSetup` method available to them so you can at least zero. There's just no - "post step" methods that run at the end for aggregating but as long as you don't need that you should be good.

std::map is definitely slower than std::vector but so is every other container type. It's not going to kill you or anything, we use std::maps all over the place. In the case of a kernel, it's perfectly parallelized in that each processor is going to just build part of the map.

If it really does show up in timing, you can always use std::unordered_map which can be faster especially if you don't insert into it very often. It's lightning fast for lookup.

WilkAndy replied:
I think i get it, thanks @permcody.

a

permcody commented on modules/porous_flow/src/userobjects/PorousFlowCount.C:
```diff
@@ -0,0 +1,108 @@
+  // before execute()
+  _num_upwinds.resize(_mesh.nElem());
+  _num_downwinds.resize(_mesh.nElem());
+  _proto_flux.resize(_mesh.nElem());
```
You definitely don't want to do this. First each processor only has a fraction of the total number of elements so it's already overestimating by potentially a lot. Assume that mesh IDs aren't continuous and just use a map instead.

permcody commented on modules/porous_flow/src/userobjects/PorousFlowCount.C:
```diff
@@ -0,0 +1,108 @@
+  // due to the execute_on = timestep_begin option,
+  // for each element.
+  // the following resize is in preparation for setProtoFlux
+  _num_upwinds[_current_elem->id()].resize(0);
```
Keep in mind that execute on an elemental Aux variable is called per quadrature point so this will get called several times. This isn't a problem with this logic, but if you do something more expensive it might be a little repetitive.

permcody commented on modules/porous_flow/include/userobjects/PorousFlowCount.h:
```diff
@@ -0,0 +1,43 @@
+protected:
+  std::vector<std::vector<std::vector<unsigned>>> _num_upwinds;
+  std::vector<std::vector<std::vector<unsigned>>> _num_downwinds;
+  std::vector<std::vector<std::vector<Real>>> _proto_flux;
```
I think I'd just go ahead and store these data structures right in your kernel. If you don't need to aggregate that information (`threadJoin()` and `finalize()`) but I don't think that's a problem for your use.

moosebuild commented:
Job `Moose Documentation:linux-clang` on 4bd539d wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9256/site

moosebuild commented:
Job `modules debug:linux-gnu` on 4bd539d : invalidated by @permcody

Several phase-field timeouts

WilkAndy commented:
This isn't quite ready for inspection yet

permcody commented on modules/porous_flow/src/kernels/PorousFlowDarcyBase.C:
```diff
@@ -120,27 +131,53 @@ PorousFlowDarcyBase::upwind(JacRes res_or_jac, unsigned int jvar)
+    const unsigned elem = _current_elem->id();
+    if (_num_upwinds.find(elem) == _num_upwinds.end())
+    {
+      _num_upwinds[elem] = std::vector<std::vector<unsigned>>(_num_phases);
```
New code paradigms to think about:
http://www.cplusplus.com/reference/unordered_map/unordered_map/emplace_hint/

This innocent line of code is actually doing a second lookup again since you just threw away the first lookup. It's not big deal, but I like to share fancy ways of doing things.

WilkAndy replied:
emplace_hint - i see, thanks.

Yes, this PR doesn't actually do anything with num_upwinds, etc, yet.  But i'm potentially concerned by the multi-threading stuff.  I don't think the multi-processor issue will be important, will it?  I just need num_upwinds and num_downwinds for _current_elem->id() to be correct.  I'm never going to look at another element while the residual is working on current_elem.

permcody reviewed (commented):
So where are you actually using this new data (_num_upwinds)? To be clear, without aggregation you'll be creating several separate data structures (one per processor, per thread) so these counts may not be accurate. I'd suggest you try this on multiple processors (and maybe multiple threads) and make sure you are happy with the results.

moosebuild commented:
Job `Moose Documentation:linux-clang` on aeeff85 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9256/site

WilkAndy commented:
So hopefully you can look at PorousFlowDarcyBase.C, @permcody , and see whether the threading or multiprocessor issues might be OK.

permcody commented:
@WilkAndy - Since you are only using these counters for a single timestep they should be just fine with MPI decomposition. The should also be fine for static thread scheduling which we do for both OpenMP and Pthread models, there's a potential for them to report the wrong counts with TBB since  that particular thread model can theoretically dynamically rebalance the workload (which means that elements that run on one thread for a particular linear iteration could potentially run on a different thread on a subsequent iteration).

So the prognosis is pretty good but it's not 100% guaranteed to work in one situation. Still it's probably "good enough" for normal use. I might suggest that we document the potential failure mode (TBB) and perhaps even print a warning when running with TBB threads. There's a compile time define `LIBMESH_HAVE_TBB_API` that you could use with I believe we have a compile time constant you could use in an if statement in conjunction with checking whether threading is actually being used `(libMesh::n_threads() > 1)`.

WilkAndy commented:
Thanks for those comments, @permcody .   I agree that it's "good enough" with the warning.  We currently believe these counters are not critical, so even if they're wrong by 1 or 2 we think it will make negligible difference to the result.  The physics is not being changed - it's just a minor tweaking to the discretisation scheme.

WilkAndy committed:
PorousFlowDarcyBase now includes counters that record, for each
node in each element, the number of times it is an upwind node
and the number of times it is a downwind node.  The counters
are incremented every nonlinear iteration.  From this info
the maximum number of upwind-downwind swaps in each element can
be computed.  If, for any timestep, the number of swaps exceeds
a user-input quantity, then the upwinding scheme is changed for
that element at that timestep.

The point of changing the upwinding scheme is to improve
convergence, since with full upwinding if the upwind direction
is changing every nonlinear iteration the Jacobian will be
poor and the nonlinear convergence will be poor.

Currently there are two alternate schemes offered.
(1) "quick" in which the nodal mobility is used, just like
full upwinding, but no mass preservation is gauranteed.
(2) "harmonic" in which the element is assigned the harmonic
mean of the nodal mobilities.  This conserves fluid mass
but is typically slower than "quick".

In all these upwinding schemes, a crucial property is that
no fluid should be withdrawn from a node where the mobility
is zero.

Fixes #8901

moosebuild commented:
Job `Moose Documentation:linux-clang` on ea32504 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9256/site

WilkAndy committed:
tests of alternative upwinding in PorousFlowDarcyBase.  Fixes #8901

moosebuild commented:
Job `Moose Documentation:linux-clang` on de61036 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9256/site

permcody commented:
This should be OK for your limited usage.

brianmoose commented:
FYI, there is a falcon test that is exodiffing due to this.

permcody commented:
@WilkAndy I merged this PR not noticing that it broke Falcon (The porous media application that was built here and partially maintained by @yidongxiainl). It's been painting our master branch orange all day now since I've merged this and I don't see what I might have broken. Do you have a minute to think about or perhaps look at this failure? It's a pretty small diff: Maybe Yidong can chime in as well. Any suggestions of fixing, skipping, or updating the test would be appreciated so we can restore the beautiful green color to our build page:
https://www.moosebuild.org/job/94390/

yidongxiainl commented:
@permcody Maybe this is the problem? "PorousFlowInternalEnergyIdeal is deprecated. Please use the IdealGasFluidProptertiesPT UserObject in the fluid_properties module. See porous_flow/tests/fluids/ideal_gas.i for an example"

permcody commented:
Well the fact that it's deprecated is not technically the problem. Even though it's deprecated it "worked" up until I merged this PR. However, if it's not a big deal to you perhaps this test could be reworked or removed. It's your application, what do you think?

yidongxiainl commented:
The failed test case is a fully coupled Thermo-Hydro-Mechanical problem. One option is I can really move it under Andy's PorousFlow module, which may still get failed. In this test, I have used all features from PorousFlow, nothing from Falcon itself. But for quick fix, I can disable the check for this test case, and make the check green. OK?

yidongxiainl commented:
BTW, what is the flag for disable the check of the case in file "tests"?

permcody commented:
`skip = 'reason'`

Usually I open a ticket and reference the ticket as the reason so I remember to eventually fix it!

yidongxiainl commented:
Eventually this test will be restored because this is a comprehensive test that involves all features of PorousFlow. I do not think currently under PorousFlow module there is such a test.

OK, cody. I will open an issue to do that.

yidongxiainl commented:
OK, I updated the devel branch of Falcon.

WilkAndy commented:
I think the "deprecated" stuff is not the problem (although it could be fixed, of course).  I think this PR actually did "break" the simulation.  This PR was exactly designed to "break" these kinds of steadystate simulations, so i'm not surprised about this exodiff (in fact i'm quite pleased to see it occurring!).

I've written "break", but really what i mean is **fix**.  Before this PR, certain kinds of PorousFlow simulations (those close to steady-state, like this one) used a different upwinding scheme than is introduced in the current PR.  I judge the upwinding schemes in this PR to be far superior, and that is the point of this PR.

The differences in solution should be minor - and they're all less than the errors brought about by any upwinding and the finite-element discretisation.  You can see this in the exodiff of the current simulation.  Stress is around 1E7 in this example, and the diffs are appearing at 1E-1.  Displacement is around 1E0 in this example, and the diffs are appearing at 1E-6.  The original gold version wouldn't have been accurate to this level anyway, because of the inferior upwinding used to generate that version.

My suggestion is to re-gold.

a

yidongxiainl commented:
@WilkAndy I will re-gold for this test soon.
